### PR TITLE
Added completion caching

### LINKED ISSUE #13: Completion for jedi isn't showing
Continued from https://github.com/Shougo/deoplete.nvim/issues/197

``python-client`` is ``neovim-0.1.3``.

[log_68422.txt](https://github.com/zchee/deoplete-jedi/files/156703/log_68422.txt)

```
# freebsd-version -ku; uname -apKU                            
11.0-CURRENT
11.0-CURRENT
FreeBSD z600 11.0-CURRENT FreeBSD 11.0-CURRENT #56: Tue Mar  1 18:52:01 CST 2016     root@z600:/usr/obj/usr/src/sys/MYKERNEL  amd64 amd64 1100100 1100099

# nvim --version
NVIM 0.1.2
Build type: Release
Compilation: /usr/bin/cc -O2 -pipe -fstack-protector -fno-strict-aliasing -Wconversion -O2 -pipe -fstack-protector -fno-strict-aliasing -DDISABLE_LOG -Wall -We
xtra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -I/wrkdirs/usr/ports/editors/neovim/work/neovim-b33c777/config -I/wrkdirs/usr/ports/editors/neovim/work/neovim-b33c777/src -I/usr/local/include -I/usr/local/include -I/usr/local/include/luajit-2.0 -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/local/include -I/usr/include -I/wrkdirs/usr/ports/editors/neovim/work/neovim-b33c777/src/nvim/auto -I/wrkdirs/usr/ports/editors/neovim/work/neovim-b33c777/include                                  Compiled by root@head-amd64-default-job-19

Optional features included (+) or not (-): +acl   +iconv    -jemalloc 
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/local/share/nvim"

```

minimal vim config:

```viml
" Your minimal init.vim

set runtimepath+=~/.vim/bundle/deoplete.nvim/
set runtimepath+=~/.vim/bundle/deoplete-jedi/

" deoplete.vim
" credit: https://gist.github.com/zchee/c314e63ae8b6bea50bb4
let g:deoplete#enable_at_startup = 1
set completeopt+=noinsert
set completeopt-=preview
let g:deoplete#enable_ignore_case = 'ignorecase'
let g:deoplete#auto_completion_start_length = 0
let g:min_pattern_length = 0
" https://github.com/Shougo/deoplete.nvim/issues/117

let g:deoplete#ignore_sources = {}
let g:deoplete#ignore_sources._ = ['buffer', 'vim', 'member']
let g:deoplete#sources#go = 'vim-go'

inoremap <expr><C-n> deoplete#mappings#manual_complete()

if exists(':DeopleteEnable')
  let g:jedi#completions_enabled = 0
  let g:jedi#auto_vim_configuration = 0
  let g:jedi#smart_auto_mappings = 0
  let g:jedi#show_call_signatures = 0
endif
```

### PR DESCRIPTION
**Pull request by tweekmonster**
Fixes #13

### CODE REVIEW

**Comment by zchee** [comment]
@tweekmonster Thanks for pull request :)
I read the code, caching behavior was able to roughly grasp.

I will test and debug from now in order to fully understand.

===

Off topic: 
Your `Dockerfile` was very helpful to me!

Previously, I've created a Dockerfile similar to it for the test method. but does not register to neovim RPC before interactive `docker run`.
I troubled a little bit, it had been allowed to stand :(
It is likely to also be applied to other plugins. Thanks!

**Comment by zchee** [comment]
> I looked through Jedi's cache source and docs a few days ago. It looks like the cache is validated after a short period. I don't think site-packages modules need to be revalidated at all. I'm not sure if that's the problem or if it's the mixed virtualenvs. From my tests, it seems like it's not using a cache at all. I can test some more and verify if using a system installed Python version produces the same results as in the Docker image.

Yes, I think also, too.
Cache file remains on the disk, but they seem to only be used for a short period.

> My other motivation for caching is because deoplete.nvim is async. deoplete-jedi calls Jedi and then formats the results. If the formatted results are cached, there will be less overhead for completion results.

Also, that is correct.
I was the first to create this source, but I did not forgive wasted long pop-up compared to the other completion engine. e.g. [zchee/deoplete-go](https://github.com/zchee/deoplete-go) and [zchee/deoplete-clang](https://github.com/zchee/deoplete-clang)
But I'm understanding to it was reduce performance.

> Another possibility with keeping a local cache is returning stale results immediately and updating the cache in a thread if a project file has been changed. That way, you get something instead of a delay.

It is very wonderful implementation!
I also use memory(thread?) cache similar to it on [zchee/deoplete-clang](https://github.com/zchee/deoplete-clang). but I did not think of trying to adapt to Python.
libclang-python also to have its implementation(but C ast), it would be effective for the optimization.

**Comment by zchee** [comment]
I test it a little, very faster than `origin/master` :) and It will work without problems.
I wait for ready to the merge.

**Comment by tweekmonster** [comment]
> It is likely to also be applied to other plugins. Thanks!

Thanks :grinning:  I made this a little while ago for testing my plugin: https://github.com/tweekmonster/vim-testbed
You can see the `.travis.yml` file using it in: https://github.com/tweekmonster/braceless.vim

I haven't had much time for it, but I plan to add support for neovim.  The main goal is for automated tests, but it could also be used for setting up baseline test environments for issue resolution.

> I also use memory(thread?) cache for completion result similar to it on zchee/deoplete-clang. but I did not think of trying to adapt to Python.

I think translation got mixed up.  I would use a memory cache for fast results, but if a buffer is changed the cache needs to be updated.  Instead of delaying to return new results, return the old results immediately and start a worker thread to update the cache with new results.

I might not add threading to this PR since I don't want to create more problems before we resolve this one :ghost:

**Comment by tweekmonster** [comment]
@zchee I added caching for local variables.  I want to test it more before submitting an official PR.

**Comment by zchee** [comment]
@tweekmonster Thanks!! I'll test later :)

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 1]
```diff
@@ -47,21 +57,120 @@ def get_complete_position(self, context):
         m = re.search(r'\w*$', context['input'])
         return m.start() if m else -1
 
+    def indent_bounds(self, line, source):
+        """Gets the bounds for a Python block.
```
Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: D401: First line should be in imperative mood ('Get', not 'Gets') by `flake8`

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 2]
```diff
@@ -47,21 +57,120 @@ def get_complete_position(self, context):
...
+                # length, we would end up too many cache entries with
+                # duplicates.
+                #
+                # Another reason is deoplete.nvim does fuzzy matching.  Without
+                # this, completing `oseo` would give no results from Jedi.
+                # With the partially cached `os` entry, it will be able to
+                # match `OSError`.
+                word = context.get('complete_str')
+                cache_key = '{}.{}'.format(buf.name,
+                                           word[:self.complete_min_length].lower())
```
Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: E501: line too long (83 > 79 characters) by `flake8`

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 3]
```diff
@@ -47,21 +57,120 @@ def get_complete_position(self, context):
...
+                        and all([filename in modules for filename in
+                                 extra_modules]) \
+                        and all([int(os.path.getmtime(filename)) == mtime
+                                 for filename, mtime in modules.items()]):
+                    return cached.get('completions', [])
+
         try:
             completions = \
-                jedi.Script(source, line, col, buf.name).completions()
+                jedi.Script('\n'.join(source), line, col, buf.name).completions()
```
Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: E501: line too long (81 > 79 characters) by flake8

**Comment by zchee** [comment]
At first, auto comments by local `flake8` settings.
but, its small problem. I did not want to remedied.

File: rplugin/python3/deoplete/sources/deoplete_jedi.py (hunk 1 above)
Comment (tweekmonster) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: flake8さん、あなたは私の英語を訂正してみろ！  But, flake8 is right :hear_no_evil:  I've never seen this warning before, though.

File: rplugin/python3/deoplete/sources/deoplete_jedi.py (hunk 2 above)
Comment (tweekmonster) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: I always ignore E501 unless it's a very long line and use my instinct instead.  I didn't break the lines because I felt 4 extra characters shouldn't cause an extra line to be created, and I didn't like the alternatives:

I don't like assignments starting with a line continuation:
```python
cache_key = \
    '{}.{}'.format(buf.name,
                   word[:self.complete_min_length].lower())
```

I don't like it when attribute/property access starts a line:
```python
cache_key = '{}.{}' \
            .format(buf.name,
                    word[:self.complete_min_length].lower())
```

and:
```python
cache_key = '{}.{}'.format(buf.name,
                           word[:self.complete_min_length]
                           .lower())
```

The last example is the most acceptable, but it's still only 4 characters over.  If you prefer strict adherence to the default flake8 rules, I'll fix it.

**Comment by tweekmonster** [comment]
Thanks, I added long comment for the first E501 warning since you don't normally code in Python.

**Comment by zchee** [comment]
> flake8さん、あなたは私の英語を訂正してみろ！

lol.
I commented 3 suggestion is a thing of flake8, not me. I do not mind.

> The last example is the most acceptable, but it's still only 4 characters over. If you prefer strict adherence to the default flake8 rules, I'll fix it.
.
.
>...you don't normally code in Python.

Yes, I usually do not write Python. 
I do not know the modern way of writing of the Python community... Therefore, I rely upon the only flake8.
I agree with your thought.

===

BTW, If this pull request is ready to merge, Could you modify pull request title?
`[WIP] Added completion caching - Don't merge` -> `Added completion caching`
I know, but users who saw would be confusing. Sorry for nit-pick :tired_face:

**Comment by zchee** [comment]
@tweekmonster 
> I haven't had much time for it, but I plan to add support for neovim. The main goal is for automated tests, but it could also be used for setting up baseline test environments for issue resolution.

I read it, and then learn it. Thanks!

**Comment by tweekmonster** [comment]
I'll update the README.md and squash the commits first.

**Comment by tweekmonster** [comment]
@zchee PR is updated and ready

**Comment by zchee** [comment]
@tweekmonster Thanks for the update and add a description to README.md!
I will final check, and merge.

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 4]
```diff
@@ -47,21 +57,120 @@ def get_complete_position(self, context):
...
+                # Typing `ose` will return only return matches that start with
+                # `ose.*`.  Typing `os` will return matches that start with
+                # `os.*`, which includes `ose.*`.  Without forcing the minimum
+                # length, we would end up too many cache entries with
+                # duplicates.
+                #
+                # Another reason is deoplete.nvim does fuzzy matching.  Without
+                # this, completing `oseo` would give no results from Jedi.
+                # With the partially cached `os` entry, it will be able to
+                # match `OSError`.
```
Comment (tweekmonster) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: I just looked closer and it seems like I'm incorrect about Jedi filtering the results.  Taking another look, but it can be a separate PR if you want to close the issue.

**Comment by tweekmonster** [comment]
I updated the script.  I got confused about Jedi's completion of scoped variables.

```python
line1 = 1

line3 = 3
```

If you're on line 2 and start a word completion, Jedi will only return `line1` as a completion.  If you're on line 4, Jedi will return `line1` and `line3`.  I misunderstood the results and assumed they were pre-filtered.

After you merge, I'm going to work on a new PR for fixing the `abbr` and `kind` values so that they're more correct.

**Comment by zchee** [comment]
@tweekmonster Oh, I'll check your said situation.

> After you merge, I'm going to work on a new PR for fixing the abbr and kind values so that they're more correct.

Thanks!

In Japan, now AM 6:00 ☀️ 
I'll merge later. just a moments :)

File: rplugin/python3/deoplete/sources/deoplete_jedi.py [hunk 5]
```diff
@@ -47,21 +57,108 @@ def get_complete_position(self, context):
...
+                lines = cached.get('lines', [0, 0])
+                modules = cached.get('modules')
+                # TODO: If cache is invalid, return stale results and start
+                # a thread to refresh.
+                if cache_line >= lines[0] and cache_line <= lines[1] \
+                        and all([filename in modules for filename in
+                                 extra_modules]) \
+                        and all([int(os.path.getmtime(filename)) == mtime
+                                 for filename, mtime in modules.items()]):
+                    self.debug('cache hit')
```
Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: I found a small bug :)
If `g:deoplete#enable_debug` is 0, will occur `[deoplete] not in debug mode, but debug called`.
https://github.com/Shougo/deoplete.nvim/blob/master/rplugin/python3/deoplete/util.py#L64-L68

and originally, for `deoplete.nvim` core.
So needs add `deoplete#sources#jedi#debug` env or delete each `self.debug` lines.

WDYT?

Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: I discuss `debug` API with @Shougo 
It does not provide now order to completion speed.

and If increase `deoplete#sources#jedi#debug` var, also need `self.vim.vars` check.
It little reduces performance. So I think better to comment-out.

WDYT?

Comment (tweekmonster) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: I forgot I left those in.  I'll remove.

> WDYT?

I think since deoplete.nvim is driven with Python, it should use python's logging module to log to a file.  Checking Vim's message log can be frustrating.

Comment (zchee) [inline rplugin/python3/deoplete/sources/deoplete_jedi.py]: @tweekmonster Thanks :)

Previously, I and Shougo considered support of python's debug logging.
Will consider again.

**Comment by zchee** [comment]
@tweekmonster 
Thanks for the fix.
LGTM.
